### ISSUE TITLE
custom/user hyphenation pattern files

### ISSUE THREAD

**Issue by jarnowic** [typst/typst#5223]
### Description

How feasible is to implement access to a user's hyphenation pattern file, locally maintained? 

### Use Case

I fear trying to integrate _every_ language into the executable may become impractical, and there's still room for custom/personalized applications of the language choice: I typeset my drafts in a mixture of greek/latin, thanks to the divergence of the alphabets in the unicode blocks.

**Comment by eltos** [comment]
This can be done by creating a show rule to insert soft hyphens at the appropriate locations:

```typ
#let words = (
  "abc-def-g",
  "foo-baz"
)
#show: it => words.fold(it, (it, word) => {
  show word.replace("-", ""): word.replace("-", sym.hyph.soft)
  it
})
```

Then
```typ
#set page(width: 105pt)
#set par(justify: true)

Hyphenating abcdefg like the alphabet.
```
Produces the lower instead of upper result
![example output](https://github.com/user-attachments/assets/c2fab7b6-ce02-4a54-9fdc-6a91e4b2bc18)



This is useful if you have special terms which are not hyphenated correctly. Not sure how performant it is if `words` is an excessive list from a file.

**Comment by laurmaedje** [comment]
A large list would likely cause considerable slowdown. In that case it would be better to merge all words into one regex so that it can all be part of one rule.

**Comment by Andrew15-5** [comment]
https://github.com/typst/typst/issues/5624

**Comment by jarnowic** [comment]
Hello! No: I didn't mean some exception mechanism, but a more general approach. In other words, some way to allow typst to load pattern files along the lines of the Knuth-Liang hyphenation algorithm dynamically, something that groff has already adopted, for instance. That way, many existing hyphenation patterns would be readily available, and some more may me developed and used following that trail. Of course, some sort of exception lists would be necessary, but the Knuth-Liang approach already takes those into account.

**Comment by Andrew15-5** [comment]
Actually, combining the approach above with reading rules from a file will result in a "good enough" approach. This can probably be a package even. 
1. read rules from user file
2. preprocess the file if necessary
3. merge the rule into a single regex pattern
4. apply a single text show rule

The API can be as simple as `show: hyphenate-rules("file.txt")`. Or like

```typ
#let (pattern, mapping) = hyphenate-rules("file.txt")
#show pattern: mapping
```

**Comment by jarnowic** [comment]
afaik, the Knuth-Liang algorithm is based on one version of "patterns", though they are not REs but a cleverly arranged pattern grouping by precedence. There are, to the best of my knowledge, at least two implementations: one in Pascal in the source code of TeX (`tex.web`, and Appendix H of _The TeXbook_); and another one in C, in the source code of groff, which adopted the same algorithm, and thus avoided duplicated patterns.

**Comment by laurmaedje** [comment]
Indeed. I wrote a bit about how it works here: https://laurmaedje.github.io/posts/hypher/

**Comment by lodwicz** [comment]
I fear I may be asking for a reversion, or at least a mixed approach; namely: allow for dynamic, run-time loading of hyphenation patterns, just as in LuaTeX or groff, instead of the static, built-in approach, proper to TeX82 and its extensions (PDFTeX or XeTeX) in their format files. In fact, it seems to me that Typst has embraced the "static" model even further, embedding the hyphenation patterns into the executable itself, while I am wondering how to take the former approach, loading patterns on demand...

**Comment by matj1** [comment]
I would like to be able to set not just patterns, but also other hyphenation rules like whether an original hyphen is retained on the beginning of the next line when a word is hyphenated at that hyphen. Different languages have different behavior in this regard, but it can't be set alone. It is the difference between languages in the following example, but I want to set it regardless of the language.

```typst
#set par(justify: true)
#set page(width: 5cm, height: 5cm, margin: 2mm)

#set text(lang: "en")
English: \
Fugiat corrupti quam sed de-bitis nulla dolores et perspi-ciatis.

#set text(lang: "cs")
Czech: \
Fugiat corrupti quam sed de-bitis nulla dolores et perspi-ciatis.
```

![rendering of the code above](https://github.com/user-attachments/assets/b983ffc7-e722-481b-8319-cc137a028f66)

**Comment by linux-user36** [comment]
@matj1 This specific behavior can already be achieved with a bit of Unicode trickery:

```typst
// A soft hyphen followed by a no-break hyphen:
#show "-": "\u{ad}\u{2011}"
``` 

Though I do agree a function akin to TeX's `\discretionary{pre-break}{post-break}{no-break}` would be handy, especially for things like pre-1996 German re-spelling of hyphenated "ck" (e.g. "gucken" → "guk-ken"). To force the arguments to be "simple", they might be restricted to strings instead of arbitrary content.
